### PR TITLE
Support setting a VersionTemplate

### PR DESCRIPTION
**Pull request by slewis74**
# Background

The current mechanics of OctoVersion suit the use case for Octopus Server, in the way it versions support releases etc, but is problematic when it comes to libraries/components. In those use cases something more "traditionally SemVer" is expected by consumers. At the moment OctoVersion is fixed in its thinking about version format.

# Result

This PR looks to add support for a `VersionTemplate`, to allow finer control for different use cases.

Given the majority use case for this will be our library/component repositories, the default template is as follows

```
{major}.{minor}.{patch}-{preReleaseTag}.{build}
```
This results in something that those used to SemVer would recognise. It also illustrates all of the possible tokens that can be used in the template.

To get the current result that Octopus Server uses, the following template can be used
```
{major}.{minor}.{build}-{preReleaseTag}
```

We could reverse this default if desired, this draft was aimed at less config for assumed majority new cases vs exact backward compat for assumed lesser cases.

## Constraints and cases considered

The contraints we are working in are:
- the build may use the OctoVersion result to tag the repo (like Octopus server does)
- tagging the repo may result in a build that calls OctoVersion (like the libraries/components will do)
- a tagged build defers the version to the tag
- a nightly build must produce a unique version number
- rebasing a branch must produce a unique version number

The uniqueness is why the build number is passed in (this is currently done via `OCTOVERSION_Patch` but would be done via `OCTOVERSION_Build` after this change).

## Nightly builds

To configure a nightly build, work on top of whichever template you already have in place. Beyond that, set the `PreReleaseTag` (`env:OCTOVERSION_PreReleaseTag`) to something like `nightly`. The string isn't important, whatever makes sense to you, it simply forces that there is a pre-release tag with that text.

### CODE REVIEW

**Comment by ryangribble** [comment]
Does this kind of break/stretch the semver notion a bit?  Eg semver doesnt actually know about a 4th thing called "build number"?  And it explicitly says not to compare build metadata when comparing versions.  So adding this 4th concept of a build number AND using that in the `CompareTo` comparisons/calculations... is it stretching the semver concept a bit too much?

To achieve what you want couldn't you just pass in a pre release tag of `nightly.<buildcounter>` ?   

Or perhaps another way to do it would be to have a `PreReleaseTag` and `PreReleaseTagSuffix` that get appended with a period separator and only when both have a value.  This wouldn't "stretch" the semver concept by introducing a new 4th variable, it more so just adds some finesse to how a pre release tag is constructed 🤔 

> To get the current result that Octopus Server uses, the following template can be used
> {major}.{minor}.{build}-{preReleaseTag}

This also kinda sounds weird.  Obviously existing usages can continue to use the default template since they pass the build counter as `Patch` but then to be more correct they could/should be enhanced to pass the actual `Build` value but then would have to override the default template to `{Major}.Minor}.{Build}` and ignore the `Patch` value... this then means there are 2 ways to achieve the same result, which feels like it leads to confusion/inconsistency.

**Comment by ryangribble** [comment]
To clarify, I think the desire to have more control over version numbers and to suit more situations such as nightly builds and other things... I'm just not sure we should create a new variable/extension to semver to achieve that...

File: source/OctoVersion.Core/Configuration/AppSettings.cs [hunk 1]
```diff
@@ -37,6 +38,8 @@ public class AppSettings : IAppSettings, IValidatableObject
 
         public string OutputJsonFile { get; set; }
 
+        public string VersionTemplate { get; set; } = "{major}.{minor}.{patch}-{preReleaseTag}.{build}";
```
Comment (uglybugger) [inline source/OctoVersion.Core/Configuration/AppSettings.cs:41]: This shouldn't know anything about our internal use cases - it should be structured according to normal SemVer.

```suggestion
        public string VersionTemplate { get; set; } = "{major}.{minor}.{patch}-{preReleaseTag}.{metadata}";
```

File: source/OctoVersion.Core/OctoVersionInfo.cs [hunk 2]
```diff
@@ -1,44 +1,50 @@
...
 using System.Text.RegularExpressions;
+using OctoVersion.Core.VersionTemplates;
 
 namespace OctoVersion.Core
 {
     public class OctoVersionInfo : SemanticVersion
     {
-        //https://semver.org/spec/v2.0.0.html#spec-item-9
-        static readonly Regex InvalidPreReleaseCharacters = new Regex("[^0-9A-Za-z-\\.]", RegexOptions.Compiled);
+        readonly VersionParser _versionParser;
```
Comment (uglybugger) [inline source/OctoVersion.Core/OctoVersionInfo.cs:9]: This class is supposed to be the "answer" class. Why does it need to take a dependency on a service?

File: source/OctoVersion.Core/OctoVersionInfo.cs [hunk 3]
```diff
@@ -1,44 +1,50 @@
...
 
         //https://semver.org/spec/v2.0.0.html#spec-item-10
         static readonly Regex InvalidBuildMetadataCharacters = new Regex("[^0-9A-Za-z-\\.]", RegexOptions.Compiled);
 
-        public OctoVersionInfo(
-            int major,
+        internal OctoVersionInfo(int major,
             int minor,
-            int patch,
+            int? patch,
```
Comment (uglybugger) [inline source/OctoVersion.Core/OctoVersionInfo.cs:16]: Patch numbers are not optional in SemVer.

File: source/OctoVersion.Core/VersionTemplates/VersionParser.cs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            if (index < 0 || index >= values.Length)
+                return string.Empty;
+            return values[index];
+        }
+
+        public SimpleVersion? TryParseSimpleVersion(string versionString)
+        {
+            try
+            {
+                if (versionString.StartsWith("v"))
```
Comment (uglybugger) [inline source/OctoVersion.Core/VersionTemplates/VersionParser.cs:139]: I don't think we should know about magic tags here. It's either semver or it's invalid.

File: source/OctoVersion.Core/VersionNumberCalculation/SimpleVersion.cs [hunk 5]
```diff
@@ -1,65 +1,36 @@
 ﻿using System;
-using System.Linq;
 
 namespace OctoVersion.Core.VersionNumberCalculation
 {
     public class SimpleVersion : IComparable
     {
-        public SimpleVersion(int major, int minor, int patch)
+        public SimpleVersion(int major, int minor, int? patch, int? build)
```
Comment (uglybugger) [inline source/OctoVersion.Core/VersionNumberCalculation/SimpleVersion.cs:7]: Patch numbers are not optional.

File: source/OctoVersion.Core/VersionNumberCalculation/SimpleVersion.cs [hunk 6]
```diff
@@ -1,65 +1,36 @@
...
         {
             Major = major;
             Minor = minor;
-            Patch = patch;
+            Patch = patch.GetValueOrDefault();
+            Build = build.GetValueOrDefault();
```
Comment (uglybugger) [inline source/OctoVersion.Core/VersionNumberCalculation/SimpleVersion.cs:12]: Why does this need to know about builds? It's calculated solely from the Git revision graph.

File: source/OctoVersion.Core/VersionNumberCalculation/SimpleVersion.cs [hunk 7]
```diff
@@ -1,65 +1,36 @@
...
         }
 
         public int Major { get; }
         public int Minor { get; }
         public int Patch { get; }
 
+        public int Build { get; }
```
Comment (uglybugger) [inline source/OctoVersion.Core/VersionNumberCalculation/SimpleVersion.cs:19]: Why does this need to know about builds? It's calculated solely from the Git revision graph.

File: source/OctoVersion.Core/VersionNumberCalculation/VersionCalculator.cs [hunk 8]
```diff
@@ -62,17 +62,24 @@ SimpleVersion GetVersionInternal(SimpleCommit commit)
...
-                version = new SimpleVersion(maxParentVersion.Major + 1, 0, 0);
+                version = new SimpleVersion(maxParentVersion.Major + 1, 0, 0, maxParentVersion.Build + 1);
             else if (commit.BumpsMinorVersion)
-                version = new SimpleVersion(maxParentVersion.Major, maxParentVersion.Minor + 1, 0);
+                version = new SimpleVersion(maxParentVersion.Major, maxParentVersion.Minor + 1, 0, maxParentVersion.Build + 1);
             else
+            {
+                // patch bumps if the direct parent is tagged
+                var parentIsTagged = commit.Parents.Any(p => p.TaggedWithVersions.Any());
+                var patch = parentIsTagged ? maxParentVersion.Patch + 1 : maxParentVersion.Patch;
```
Comment (uglybugger) [inline source/OctoVersion.Core/VersionNumberCalculation/VersionCalculator.cs:76]: I think this is distorting thing way too much for internal-specific use cases.

File: source/OctoVersion.Core/SemanticVersion.cs [hunk 9]
```diff
@@ -1,101 +1,47 @@
...
-        }
-
         public int CompareTo(object obj)
         {
             if (!(obj is SemanticVersion other)) return -1;
 
             if (Major != other.Major) return Major.CompareTo(other.Major);
             if (Minor != other.Minor) return Minor.CompareTo(other.Minor);
             if (Patch != other.Patch) return Patch.CompareTo(other.Patch);
+            if (Build != other.Build) return Build.CompareTo(other.Build);
```
Comment (uglybugger) [inline source/OctoVersion.Core/SemanticVersion.cs:36]: Semantic versions don't know about builds.

**Review by uglybugger — CHANGES_REQUESTED** [review]
I'm not convinced that this is taking things in the direction we discussed.

Rather than taking it further async, I think we should arrange to pair on a refactoring of what's currently in `main`.

**Comment by slewis74** [comment]
Closing as this isn't the path we are looking for.

**Pull request closed without merging**